File: client/python/tests/test_BBBrowserExtensionInjector.py
from __future__ import annotations

import os
from pathlib import Path
import unittest

from modcdp import ModCDPClient
from modcdp.BBBrowserExtensionInjector import BBBrowserExtensionInjector

HERE = Path(__file__).resolve().parent
EXTENSION_PATH = HERE.parents[2] / "dist" / "extension"


class BBBrowserExtensionInjectorTests(unittest.TestCase):
    def test_uses_configured_extension_id(self) -> None:
        injector = BBBrowserExtensionInjector({"extension_id": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"})
        injector.prepare()
        self.assertEqual(injector.getLauncherConfig()["extension_id"], "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa")

    @unittest.skipUnless(os.environ.get("BROWSERBASE_API_KEY", "").strip(), "BROWSERBASE_API_KEY is required for live Browserbase tests")
    def test_uploads_real_extension_and_launches_browserbase_browser_with_it_installed(self) -> None:
        cdp = ModCDPClient(
            launch={
                "mode": "bb",
                "options": {
                    "project_id": os.environ.get("BROWSERBASE_PROJECT_ID"),
                    "timeout": 120,
                    **({"region": os.environ["BROWSERBASE_REGION"]} if os.environ.get("BROWSERBASE_REGION") else {}),
                },
            },
            upstream={"mode": "ws"},
            extension={
                "mode": "inject",
                "path": str(EXTENSION_PATH),
                "service_worker_url_suffixes": ["/modcdp/service_worker.js"],
                "trust_service_worker_target": True,
            },
        )

        try:
            cdp.connect()
            self.assertEqual(cdp.connect_timing["extension_source"] if cdp.connect_timing else None, "bb")
            self.assertIsInstance(cdp.extension_id, str)
            service_worker_url = cdp.Mod.evaluate(expression="chrome.runtime.getURL('modcdp/service_worker.js')")
            self.assertRegex(str(service_worker_url), r"^chrome-extension://[a-z]{32}/modcdp/service_worker\.js$")
        finally:
            cdp.close()


if __name__ == "__main__":
    unittest.main()

File: test/test.BBBrowserExtensionInjector.ts
import assert from "node:assert/strict";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { describe, it } from "vitest";

import { BBBrowserExtensionInjector } from "../bridge/BBBrowserExtensionInjector.js";
import { ModCDPClient } from "../client/js/ModCDPClient.js";

const HERE = path.dirname(fileURLToPath(import.meta.url));
const EXTENSION_PATH = path.resolve(HERE, "..", "dist", "extension");
const hasBrowserbaseEnv = Boolean(process.env.BROWSERBASE_API_KEY?.trim());

describe("BBBrowserExtensionInjector", () => {
  it("uses configured extension id", async () => {
    const injector = new BBBrowserExtensionInjector({ extension_id: "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa" });

    await injector.prepare();

    assert.equal(injector.getLauncherConfig().extension_id, "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa");
  });

  it("does not upload when no extension path or id is configured", async () => {
    const injector = new BBBrowserExtensionInjector();

    await injector.prepare();

    assert.deepEqual(injector.getLauncherConfig(), {});
    assert.equal((injector as unknown as { zip_path: string | null }).zip_path, null);
  });
});

describe.skipIf(hasBrowserbaseEnv)("BBBrowserExtensionInjector without credentials", () => {
  it("requires BROWSERBASE_API_KEY when extension upload is needed", async () => {
    const injector = new BBBrowserExtensionInjector({ extension_path: EXTENSION_PATH });

    try {
      await assert.rejects(() => injector.prepare(), /BROWSERBASE_API_KEY/);
      assert.equal((injector as unknown as { cleanup: unknown }).cleanup, null);
    } finally {
      await injector.close();
    }
  });
});

describe.skipIf(!hasBrowserbaseEnv)("BBBrowserExtensionInjector", () => {
  it(
    "uploads the real extension and launches a Browserbase browser with it installed",
    { timeout: 180_000 },
    async () => {
      const cdp = new ModCDPClient({
        launch: {
          mode: "bb",
          options: {
            project_id: process.env.BROWSERBASE_PROJECT_ID,
            timeout: 120,
            ...(process.env.BROWSERBASE_REGION ? { region: process.env.BROWSERBASE_REGION } : {}),
          },
        },
        upstream: { mode: "ws" },
        extension: {
          mode: "inject",
          path: EXTENSION_PATH,
          service_worker_url_suffixes: ["/modcdp/service_worker.js"],
          trust_service_worker_target: true,
        },
      });

      try {
        await cdp.connect();
        assert.equal(cdp.connect_timing?.extension_source, "bb");
        assert.equal(typeof cdp.extension_id, "string");
        const service_worker_url = await cdp.Mod.evaluate({
          expression: "chrome.runtime.getURL('modcdp/service_worker.js')",
        });
        assert.match(String(service_worker_url), /^chrome-extension:\/\/[a-z]{32}\/modcdp\/service_worker\.js$/);
      } finally {
        await cdp.close();
      }
    },
  );
});
